File: src/features/progression/finalize.ts
/**
 * Build the durable progression snapshot persisted after a workout (Phase 11).
 * Pure — composes the tested XP curve (applyXp) and the ranking engine (Hunter
 * score/rank). The result is handed to the guarded apply_session_progression RPC.
 *
 * Attributes: a null means "not recomputed this workout" — carry the current
 * value. For the Hunter blend, an attribute of 0 is treated as "no data yet" and
 * renormalized out, so new users get a neutral Hunter score rather than zero.
 */
import type { Rank } from '@/constants/ranks';
import { applyXp, type ProgressionSnapshot } from '@/features/progression/rewards';
import {
  disciplineScore,
  exerciseScore,
  hunterRank,
  hunterScore,
  physiqueScore,
  strengthScore,
} from '@/services/ranking';

export interface AttributeInputs {
  strength: number | null;
  physique: number | null;
  endurance: number | null;
  discipline: number | null;
}

export interface FinishInputs {
  bodyweightKg: number | null;
  sex: string | null;
  exercises: { name: string; best1RMkg: number | null }[];
  assessment: {
    bodyFatPercent: number | null;
    skeletalMuscleMassKg: number | null;
    weightKg: number | null;
    sex: string | null;
  } | null;
}

/**
 * Recompute attributes after a workout from gathered inputs (pure). Strength is
 * the average of ranked exercise scores; physique comes from the latest
 * assessment (null if none); discipline from the streak; endurance is carried.
 */
export function computeAttributes(inputs: FinishInputs, newStreakDays: number): AttributeInputs {
  const scores = inputs.exercises
    .map((e) =>
      exerciseScore({
        exerciseName: e.name,
        bestEstimated1rmKg: e.best1RMkg,
        bodyweightKg: inputs.bodyweightKg,
        sex: inputs.sex,
      }),
    )
    .filter((s): s is number => s != null);

  return {
    strength: strengthScore(scores),
    physique: physiqueScore(inputs.assessment),
    endurance: null,
    discipline: disciplineScore({ currentStreakDays: newStreakDays, adherence: 1 }),
  };
}

export interface ProgressionUpdateInput {
  current: ProgressionSnapshot;
  currentAttributes: { strength: number; physique: number; endurance: number; discipline: number };
  xpEarned: number;
  streak: { current: number; longest: number };
  /** Newly computed attributes; null keeps the current value. */
  attributes: AttributeInputs;
}

export interface ProgressionPersistPayload {
  level: number;
  current_xp: number;
  lifetime_xp: number;
  strength_score: number;
  physique_score: number;
  endurance_score: number;
  discipline_score: number;
  hunter_score: number;
  hunter_rank: Rank;
  current_streak_days: number;
  longest_streak_days: number;
}

export function buildProgressionUpdate(input: ProgressionUpdateInput): ProgressionPersistPayload {
  const xp = applyXp(input.current, input.xpEarned);

  const resolved = {
    strength: input.attributes.strength ?? input.currentAttributes.strength,
    physique: input.attributes.physique ?? input.currentAttributes.physique,
    endurance: input.attributes.endurance ?? input.currentAttributes.endurance,
    discipline: input.attributes.discipline ?? input.currentAttributes.discipline,
  };

  const meaningful = (v: number) => (v > 0 ? v : null);
  const score = hunterScore({
    strength: meaningful(resolved.strength),
    physique: meaningful(resolved.physique),
    endurance: meaningful(resolved.endurance),
    discipline: meaningful(resolved.discipline),
  });

  return {
    level: xp.level,
    current_xp: xp.currentXp,
    lifetime_xp: xp.lifetimeXp,
    strength_score: resolved.strength,
    physique_score: resolved.physique,
    endurance_score: resolved.endurance,
    discipline_score: resolved.discipline,
    hunter_score: score,
    hunter_rank: hunterRank(score),
    current_streak_days: input.streak.current,
    longest_streak_days: input.streak.longest,
  };
}

File: android-native/app/src/main/java/com/fitnessrpg/app/domain/progression/Finalize.kt
package com.fitnessrpg.app.domain.progression

import com.fitnessrpg.app.domain.rank.Rank
import com.fitnessrpg.app.domain.ranking.DisciplineInput
import com.fitnessrpg.app.domain.ranking.disciplineScore

/**
 * Build the durable progression snapshot persisted after a workout. Hunter Rank
 * now comes from the three-pillar engine (physique + strength + conditioning),
 * NOT a weighted average that includes discipline. Discipline is still computed
 * for XP/streaks/quests, but never feeds Hunter Rank.
 *
 * Physique deliberately does NOT read the ambiguous skeletal-muscle-mass column
 * (existing data may hold total muscle mass there); it uses body fat + FFMI.
 */
data class AttributeInputs(
    val strength: Double?,
    val physique: Double?,
    val endurance: Double?,
    val discipline: Double?,
)

/** Workout completion updates discipline only. Physical pillars are assessment-owned. */
fun computeWorkoutAttributes(newStreakDays: Int): AttributeInputs {
    return AttributeInputs(
        strength = null,
        physique = null,
        endurance = null,
        discipline = disciplineScore(DisciplineInput(newStreakDays, 1.0)),
    )
}

data class CurrentAttributes(
    val strength: Double,
    val physique: Double,
    val endurance: Double,
    val discipline: Double,
    val hunterScore: Double = 0.0,
    val hunterRank: Rank = Rank.E,
)

data class StreakSnapshot(val current: Int, val longest: Int)

data class ProgressionUpdateInput(
    val current: ProgressionSnapshot,
    val currentAttributes: CurrentAttributes,
    val xpEarned: Int,
    val streak: StreakSnapshot,
    val attributes: AttributeInputs,
)

data class ProgressionPersistPayload(
    val level: Int,
    val currentXp: Int,
    val lifetimeXp: Int,
    val strengthScore: Double,
    val physiqueScore: Double,
    val enduranceScore: Double,
    val disciplineScore: Double,
    val hunterScore: Double,
    val hunterRank: Rank,
    val currentStreakDays: Int,
    val longestStreakDays: Int,
)

fun buildProgressionUpdate(input: ProgressionUpdateInput): ProgressionPersistPayload {
    val xp = applyXp(input.current, input.xpEarned)

    val resolvedStrength = input.attributes.strength ?: input.currentAttributes.strength
    val resolvedPhysique = input.attributes.physique ?: input.currentAttributes.physique
    val resolvedEndurance = input.attributes.endurance ?: input.currentAttributes.endurance
    val resolvedDiscipline = input.attributes.discipline ?: input.currentAttributes.discipline

    return ProgressionPersistPayload(
        level = xp.level,
        currentXp = xp.currentXp,
        lifetimeXp = xp.lifetimeXp,
        strengthScore = resolvedStrength,
        physiqueScore = resolvedPhysique,
        enduranceScore = resolvedEndurance,
        disciplineScore = resolvedDiscipline,
        // Workout completion changes XP/streaks and raw evidence. Hunter Rank is
        // recalculated only by the assessment pipeline, never from lifetime PRs.
        hunterScore = input.currentAttributes.hunterScore,
        hunterRank = input.currentAttributes.hunterRank,
        currentStreakDays = input.streak.current,
        longestStreakDays = input.streak.longest,
    )
}
